// @suborbital/runnable/dist/modules/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = require("./base");
const env_1 = require("../bindings/env");
const decoder = new TextDecoder();
class Request extends base_1.default {
    method() {
        return decoder.decode(this.getField(env_1.FieldType.Meta, "method"));
    }
    url() {
        return decoder.decode(this.getField(env_1.FieldType.Meta, "url"));
    }
    id() {
        return decoder.decode(this.getField(env_1.FieldType.Meta, "id"));
    }
    body() {
        return decoder.decode(this.getField(env_1.FieldType.Body, "body"));
    }
    bodyBytes() {
        return this.getField(env_1.FieldType.Body, "body");
    }
    bodyField(key) {
        return decoder.decode(this.getField(env_1.FieldType.Body, key));
    }
    header(key) {
        return decoder.decode(this.getField(env_1.FieldType.Header, key));
    }
    urlParam(key) {
        return decoder.decode(this.getField(env_1.FieldType.Params, key));
    }
    state(key) {
        return decoder.decode(this.getField(env_1.FieldType.State, key));
    }
    stateBytes(key) {
        return this.getField(env_1.FieldType.State, key);
    }
    getField(field, key) {
        const resultSize = this.env.requestGetField(field, key, this.ident);
        return this.ffiResult(resultSize);
    }
}
exports.default = Request;
