// @suborbital/runnable/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setup = exports.runnable = exports.request = exports.log = exports.http = exports.graphql = exports.file = exports.cache = exports.db = void 0;
const db_1 = require("./modules/db");
const cache_1 = require("./modules/cache");
const file_1 = require("./modules/file");
const graphql_1 = require("./modules/graphql");
const http_1 = require("./modules/http");
const log_1 = require("./modules/log");
const request_1 = require("./modules/request");
const runnable_1 = require("./modules/runnable");
const ffi_1 = require("./modules/ffi");
exports.db = new db_1.default();
exports.cache = new cache_1.default();
exports.file = new file_1.default();
exports.graphql = new graphql_1.default();
exports.http = new http_1.default();
exports.log = new log_1.default();
exports.request = new request_1.default();
exports.runnable = new runnable_1.default();
function setup(imports, ident) {
    // Apply the import object
    // @ts-ignore
    ffi_1.default.env._exports = imports;
    ffi_1.default.ident = ident;
}
exports.setup = setup;
